firedrake/assemble_expressions.py

```diff
@@ -158,19 +180,25 @@ def _split_component_tensor(o, self):
 
 
 @_split.register(Indexed)
-def _split_indexed(o, self):
+def _split_indexed(o, self, inct):
     aggregate, multiindex = o.ufl_operands
     indices = multiindex.indices()
     result = []
-    for agg in self(aggregate):
+    for agg in self(aggregate, False):
         ncmp = len(agg.ufl_shape)
-        idx = indices[:ncmp]
-        indices = indices[ncmp:]
         if ncmp == 0:
             result.append(agg)
-        else:
+        elif not inct:
+            idx = indices[:ncmp]
+            indices = indices[ncmp:]
             mi = multiindex if multiindex.indices() == idx else MultiIndex(idx)
             result.append(ufl_reuse_if_untouched(o, agg, mi))
+        else:
+            # shape and inct
+            aggshape = (flatten(agg.ufl_shape)
+                        + tuple(itertools.repeat(1, len(aggregate.ufl_shape) - 1)))
```

Comment:
Do we ever get here with `len(aggregate.ufl_shape) != 1`?

Reply:
I suspect I did it for belt-and-braces.

Reply:
I guess know because `split_coefficient` flattens.